packages/test/test-end-to-end-tests/src/test/gcDataStoreRequests.spec.ts

```diff
@@ -66,51 +68,85 @@ describeNoCompat("GC Data Store Requests", (getTestObjectProvider) => {
-                resolve(true);
-            } else if (op.type === "summaryNack") {
-                reject(new Error("summaryNack"));
-            }
-        }));
-        return handle;
+    /**
+     * Wait for an op with the given key in the given shared directory.
+     * @returns the sequence number of this op.
+     */
+    async function waitForDirectoryOp(directory: ISharedDirectory, key: string): Promise<number> {
+        let sequenceNumber: number | undefined;
+        await new Promise((resolve) => {
+            const listener = (op: ISequencedDocumentMessage) => {
+                if (op.contents.key === key) {
+                    sequenceNumber = op.sequenceNumber;
+                    resolve(true);
+                    directory.off("op", listener);
+                }
+            };
+            directory.on("op", listener);
+        });
+        assert(sequenceNumber !== undefined);
+        return sequenceNumber;
     }
 
+    const createContainer = async (): Promise<IContainer> => provider.createContainer(runtimeFactory);
+    const loadContainer = async (summaryVersion: string): Promise<IContainer> => {
+        const requestHeader = {
+            [LoaderHeader.version]: summaryVersion,
+        };
+        return provider.loadContainer(runtimeFactory, undefined /* options */, requestHeader);
+    };
+
     beforeEach(async () => {
         provider = getTestObjectProvider();
 
         // Create a Container for the first client.
-        container1 = await createContainer();
+        mainContainer = await createContainer();
+
+        // Create and setup a summary collection that will be used to track and wait for summaries.
+        summaryCollection = new SummaryCollection(
+            mainContainer.deltaManager.initialSequenceNumber,
+            new TelemetryNullLogger(),
+        );
+        mainContainer.deltaManager.inbound.on("op",
+            (op) => summaryCollection.handleOp(op));
     });
 
     it("should fail requests with externalRequest flag for unreferenced data stores", async () => {
-        const dataStore1 = await requestFluidObject<TestDataObject>(container1, "default");
+        const directoryKey = "dataStore2";
+        const dataStore1 = await requestFluidObject<TestDataObject>(mainContainer, "default");
 
         // Create a second data store (dataStore2) and add its handle to mark it as referenced.
         const dataStore2 = await factory.createInstance(dataStore1._context.containerRuntime);
-        dataStore1._root.set("dataStore2", dataStore2.handle);
+        dataStore1._root.set(directoryKey, dataStore2.handle);
 
-        // Wait for ops to be processed so that summarizer creates dataStore2.
-        await provider.ensureSynchronized();
+        // Wait for the set to be processed. Then wait for a summary that includes the sequence number of the above op.
+        let sequenceNumber = await waitForDirectoryOp(dataStore1._root, directoryKey);
```

Comment:
i'd perfer if we didn't need to watch for a specific op. could you instead just make modification, then wait for the container to no longer be dirty, then grab the last sequence number off the delta manager?

Reply:
you could probably encapsulate all of the is waitForSummary, so just make any change you want, then waitForSummary waits for the next summary after !dirty